xcryptofork/patches/0002-Implement-algorithms-using-backend-proxies.patch

```diff
@@ -0,0 +1,214 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Davis Goodin <dagood@microsoft.com>
+Date: Thu, 28 Sep 2023 23:12:08 -0500
+Subject: [PATCH] Implement algorithms using backend proxies
+
+Implement:
+
+* hkdf
+* md4
+* pbkdf2
+* several sha3 hashes
+---
+ hkdf/hkdf.go     | 16 ++++++++++++++++
+ md4/md4.go       |  9 +++++++++
+ pbkdf2/pbkdf2.go |  9 +++++++++
+ sha3/hashes.go   | 27 +++++++++++++++++++++++++++
+ 4 files changed, 61 insertions(+)
+
+diff --git a/hkdf/hkdf.go b/hkdf/hkdf.go
+index dda3f143bec506..13180f5c71a24d 100644
+--- a/hkdf/hkdf.go
++++ b/hkdf/hkdf.go
+@@ -15,6 +15,8 @@ import (
+ 	"errors"
+ 	"hash"
+ 	"io"
++
++	"golang.org/x/crypto/internal/backend"
+ )
+ 
+ // Extract generates a pseudorandom key for use with Expand from an input secret
+@@ -24,6 +26,13 @@ import (
+ // Expand invocations and different context values. Most common scenarios,
+ // including the generation of multiple keys, should use New instead.
+ func Extract(hash func() hash.Hash, secret, salt []byte) []byte {
++	if backend.Enabled && backend.SupportsHKDF() {
++		key, err := backend.ExtractHKDF(hash, secret, salt)
++		if err != nil {
++			panic("x/crypto/hkdf: " + err.Error())
++		}
++		return key
++	}
+ 	if salt == nil {
+ 		salt = make([]byte, hash().Size())
+ 	}
+@@ -81,6 +90,13 @@ func (f *hkdf) Read(p []byte) (int, error) {
+ // random or pseudorandom cryptographically strong key. See RFC 5869, Section
+ // 3.3. Most common scenarios will want to use New instead.
+ func Expand(hash func() hash.Hash, pseudorandomKey, info []byte) io.Reader {
++	if backend.Enabled && backend.SupportsHKDF() {
```

Comment:
TODO: Handle MD5? Used in benchmarks, causes a panic.

On the note of benchmarks, this is what I get for hkdf as-is, ignoring md5:

```
goos: linux
goarch: amd64
pkg: golang.org/x/crypto/hkdf
cpu: Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz
                     │ nobackend.txt │               openssl.txt                │
                     │    sec/op     │     sec/op      vs base                  │
20ByteSHA1Single-8      2.558µ ± 16%     20.715µ ± 4%    +709.79% (p=0.000 n=8)
32ByteSHA256Single-8    3.136µ ± 12%     22.067µ ± 6%    +603.67% (p=0.000 n=8)
64ByteSHA512Single-8    5.144µ ± 30%     23.604µ ± 5%    +358.90% (p=0.000 n=8)
8ByteSHA1Stream-8       152.5n ±  2%    42068.0n ± 7%  +27485.57% (p=0.000 n=8)
20ByteSHA1Stream-8      359.9n ±  3%    41875.0n ± 5%  +11535.18% (p=0.000 n=8)
8ByteSHA256Stream-8     129.5n ±  3%    61891.5n ± 7%  +47674.22% (p=0.000 n=8)
32ByteSHA256Stream-8    479.4n ±  2%    62910.0n ± 5%  +13022.65% (p=0.000 n=8)
8ByteSHA512Stream-8     89.56n ±  2%   86508.00n ± 7%  +96497.62% (p=0.000 n=8)
64ByteSHA512Stream-8    629.7n ±  2%    88247.5n ± 6%  +13914.21% (p=0.000 n=8)
geomean                 584.2n            43.56µ        +7356.21%

                     │  nobackend.txt   │             openssl.txt              │
                     │       B/s        │      B/s       vs base               │
20ByteSHA1Single-8       7636.7Ki ± 13%   942.4Ki ±  4%  -87.66% (p=0.000 n=8)
32ByteSHA256Single-8      9.732Mi ± 17%   1.383Mi ±  7%  -85.79% (p=0.000 n=8)
64ByteSHA512Single-8     11.873Mi ± 22%   2.589Mi ±  4%  -78.19% (p=0.000 n=8)
8ByteSHA1Stream-8       51225.6Ki ±  4%   185.5Ki ±  5%  -99.64% (p=0.000 n=8)
20ByteSHA1Stream-8      54267.6Ki ±  5%   468.8Ki ±  8%  -99.14% (p=0.000 n=8)
8ByteSHA256Stream-8     60312.5Ki ±  5%   127.0Ki ± 15%  -99.79% (p=0.000 n=8)
32ByteSHA256Stream-8    65185.5Ki ±  3%   498.0Ki ±  4%  -99.24% (p=0.000 n=8)
8ByteSHA512Stream-8    87241.21Ki ±  3%   87.89Ki ± 11%  -99.90% (p=0.000 n=8)
64ByteSHA512Stream-8    99277.3Ki ±  4%   708.0Ki ±  8%  -99.29% (p=0.000 n=8)
geomean                   34.58Mi         474.1Ki        -98.66%
```